activerecord/lib/active_record/autosave_association.rb

```diff
@@ -323,14 +324,21 @@ def validate_collection_association(reflection)
       # Returns whether or not the association is valid and applies any errors to
       # the parent, <tt>self</tt>, if it wasn't. Skips any <tt>:autosave</tt>
       # enabled records if they're marked_for_destruction? or destroyed.
-      def association_valid?(reflection, record)
+      def association_valid?(reflection, record, index=nil)
         return true if record.destroyed? || record.marked_for_destruction?
 
         validation_context = self.validation_context unless [:create, :update].include?(self.validation_context)
         unless valid = record.valid?(validation_context)
           if reflection.options[:autosave]
             record.errors.each do |attribute, message|
-              attribute = "#{reflection.name}.#{attribute}"
+              options = self.nested_attributes_options[reflection.name] 
+              if ActiveRecord::Base.index_nested_attribute_errors 
+                attribute = "#{reflection.name}[#{index}].#{attribute}"
+              elsif index.nil? or options.nil? or not options[:index_errors]
```

Comment:
Done. We'll update once we figured out how we should break the conditional up (see above).